File: online_taksi_flutter/lib/network/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/Constants.dart';
import '../core/session_manager.dart';

class ApiClient {
  final SessionManager sessionManager;

  ApiClient({
    required this.sessionManager,
  });

  Future<Map<String, dynamic>> post(
      String path,
      Map<String, dynamic> body,
      ) async {
    final token = await sessionManager.getToken();

    final response = await http.post(
      Uri.parse(Constants.baseUrl + path),
      headers: {
        "Content-Type": "application/json",
        if (token != null && token.isNotEmpty) "Authorization": "Bearer $token",
      },
      body: jsonEncode(body),
    );

    return _handleResponse(response);
  }

  Future<Map<String, dynamic>> put(
      String path,
      Map<String, dynamic> body,
      ) async {
    final token = await sessionManager.getToken();

    final response = await http.put(
      Uri.parse(Constants.baseUrl + path),
      headers: {
        "Content-Type": "application/json",
        if (token != null && token.isNotEmpty) "Authorization": "Bearer $token",
      },
      body: jsonEncode(body),
    );

    return _handleResponse(response);
  }

  Future<Map<String, dynamic>> get(String path) async {
    final token = await sessionManager.getToken();

    final response = await http.get(
      Uri.parse(Constants.baseUrl + path),
      headers: {
        "Content-Type": "application/json",
        if (token != null && token.isNotEmpty) "Authorization": "Bearer $token",
      },
    );

    return _handleResponse(response);
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    final decoded = response.body.isNotEmpty
        ? jsonDecode(response.body) as Map<String, dynamic>
        : <String, dynamic>{};

    if (response.statusCode >= 200 && response.statusCode < 300) {
      return decoded;
    }

    throw Exception(decoded["detail"]?.toString() ?? response.body);
  }
}

File: online_taksi_react_native/src/network/apiClient.ts
import { Constants } from "../core/constants";
import { SessionManager } from "../core/sessionManager";

type HttpMethod = "GET" | "POST" | "PUT" | "DELETE";

async function request<T>(
    path: string,
    method: HttpMethod,
    body?: unknown
): Promise<T> {
    const token = await SessionManager.getToken();

    const response = await fetch(Constants.BASE_URL + path, {
        method,
        headers: {
            "Content-Type": "application/json",
            ...(token ? { Authorization: `Bearer ${token}` } : {}),
        },
        body: body ? JSON.stringify(body) : undefined,
    });

    const text = await response.text();
    const json = text ? JSON.parse(text) : {};

    if (!response.ok) {
        const message =
            typeof json?.detail === "string" ? json.detail : text || "İstek başarısız";
        throw new Error(message);
    }

    return json as T;
}

export const ApiClient = {
    get<T>(path: string) {
        return request<T>(path, "GET");
    },

    post<T>(path: string, body: unknown) {
        return request<T>(path, "POST", body);
    },

    put<T>(path: string, body?: unknown) {
        return request<T>(path, "PUT", body ?? {});
    },
};